# Proyecto: Análisis de ventas de auto por Ebay

El objetivo de este proyecto es realizar un análisis general de ventas de autos usados que se realizan mediante la página Ebay.

Los datos recopilados se pueden encontrar [Aqui](https://www.kaggle.com/orgesleka/used-cars-database/data). De ellos se extrajo una muestra de 50.000 datos y el significado de cada encabezado se puede encontrar en el mismo link. 

In [1]:
import pandas as pd, numpy as np
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

Se observa lo siguiente:
* Hay 20 columnas, la mayoria strings.
* Existen algunas columnas con valores Nulls ( pues no tienen los 50.000 datos)
* Los encabezados estan escritos DeEstaForma (camelcase), en lugar de_esta_forma (snakecase), la cual es mejor para Python.

A continuación se realizará las siguientes operaciones:
- Se va a cambiar de camelcase a snakecase.
- Se renombrara algunos encabezados por unos mas descriptivos. 


In [4]:
autos.columns  #observamos los encabezados

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
autos.columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'] #renombramos algunos encabezados

In [6]:
autos.head() #se observa como quedó

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Cleaning de data

Se explorara la data para definir que limpieza de datos es necesaria:

    i.- Contenido de las columnas donde casi todos los valores sean lo mismo ( en cuyo caso se puede eliminar la columna, por inutil)
    ii.- Ver aquellas columnas que contiene informacioón numérica  y limpiarla en caso de que sea necesario.


In [7]:
autos.describe(include='all') #solo mostrará aquellas columnas numéricas. 

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-23 19:38:20  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

Se puede concluir lo siguiente:
* Todas las columnas que tienen a lo mas un valor ( o 2) como unique, son candidatas a ser eliminadas. 
* Las columnas que tienen valores numéricos como string deben ser limpiados y transformados de formato. 

* El resto de las columnas necesitan mayor investigacion


In [8]:
autos.drop(["seller", "offer_type","nr_of_pictures"], axis=1, inplace=True) #eliminar columnas
autos['price'] = (autos['price'].str.replace('$','').str.replace(',','').str.strip().astype(int) ) # de string a int
autos['odometer'] = (autos['odometer'].str.replace('km','').str.replace(',','').str.strip().astype(int)) # de string a int

In [9]:
autos.rename({"odometer": "odometer_km"}, axis=1, inplace=True) #cambiando el nombre de la columna

Se analiza "Price" mas a fondo:

In [10]:
print("Numero de precios Unicos:",autos["price"].unique().shape) # ver cuantos valores unicos hay
print("Descripcion de los datos de Precio:",autos["price"].describe()) #busqueda de valores outlier
autos["price"].value_counts().head(20)

Numero de precios Unicos: (2357,)
Descripcion de los datos de Precio: count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64


0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
700      395
4500     394
300      384
2200     382
950      379
Name: price, dtype: int64

De lo aterior se puede concluir lo siguiente:
* la variead de precios parece correcta (2357).
* Se observa que existen 1421 precios de valor 0, lo que indica un problema con estos valores. 
* El valor maximo es de 100 millones de dores lo resulta algo poco probable, debido a esto veremos entre que valores se mueven los precios mas altos

### Analisis de "Prices"

In [11]:
autos["price"].value_counts().sort_index(ascending=False).head(20)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
220000      1
198000      1
197000      1
Name: price, dtype: int64

In [12]:
autos["price"].value_counts().sort_index(ascending=True).head(70)

0      1421
1       156
2         3
3         1
5         2
8         1
9         1
10        7
11        2
12        3
13        2
14        1
15        2
17        3
18        1
20        4
25        5
29        1
30        7
35        1
40        6
45        4
47        1
49        4
50       49
55        2
59        1
60        9
65        5
66        1
       ... 
115       2
117       1
120      39
122       1
125       8
129       1
130      15
135       1
139       1
140       9
145       2
149       7
150     224
156       2
160       8
170       7
173       1
175      12
179       1
180      35
185       1
188       1
190      16
193       1
195       2
198       1
199      41
200     266
205       1
210       1
Name: price, Length: 70, dtype: int64

Se observa que solo existe una cifra muy alta, la que se debe remover. Ahora removiendo los valores que resultan outlier ( muy pequeños  y valores demasiado altos)

In [13]:
autos = autos[autos["price"].between(49,351000)]  #sacando las filas de los outlier
autos["price"].describe()

count     48351.000000
mean       5914.970776
std        9071.407458
min          49.000000
25%        1250.000000
50%        3000.000000
75%        7499.000000
max      350000.000000
Name: price, dtype: float64

### Análisis de la columna "odometer"

autos["odometer_km"].value_counts().sort_index(ascending = False).head(20) #observando los valores mayores

In [14]:
autos["odometer_km"].value_counts().sort_index(ascending = True).head(20) #observando los valores menores

5000        784
10000       249
20000       759
30000       778
40000       815
50000      1011
60000      1153
70000      1215
80000      1414
90000      1733
100000     2106
125000     5045
150000    31289
Name: odometer_km, dtype: int64

No parece haber inconsistencias en los valores de "odometer_km"

### Análisis de la columna con valores de tipo Date

Las columnas siguientes son de tipo datos:

- `date_crawled`
- `last_seen`
- `ad_created`
- `registration_month`
- `registration_year`

Veamos de que tipo de datos lo reconoce Python


In [15]:
autos.dtypes

date_crawled          object
name                  object
price                  int64
abtest                object
vehicle_type          object
registration_year      int64
gearbox               object
power_ps               int64
model                 object
odometer_km            int64
registration_month     int64
fuel_type             object
brand                 object
unrepaired_damage     object
ad_created            object
postal_code            int64
last_seen             object
dtype: object

Para las primeras tres columnas descritas anteriormente vemos que estan reconocidas como strings (object). Por lo que se realizara su conversion a tipo de datos Date. 

Exploramos estas tres columnas.


In [16]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [17]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index().head() #str[:10], toma los 10 primeros caracteres, normalize-todo en%

2016-03-05    0.025356
2016-03-06    0.014064
2016-03-07    0.036028
2016-03-08    0.033257
2016-03-09    0.033050
Name: date_crawled, dtype: float64

In [18]:
autos["date_crawled"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending=False).head() 

2016-04-07    0.001386
2016-04-06    0.003164
2016-04-05    0.013050
2016-04-04    0.036525
2016-04-03    0.038572
Name: date_crawled, dtype: float64

Se observa uniformidad en la distribucion de las fechas y no hay maximo ni minimos que sugieran alguna anomalia.

In [19]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index().head() 

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
Name: ad_created, dtype: float64

In [20]:
autos["ad_created"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = False ).head() 

2016-04-07    0.001241
2016-04-06    0.003247
2016-04-05    0.011768
2016-04-04    0.036897
2016-04-03    0.038820
Name: ad_created, dtype: float64

Para "ad_created" obtenemos la misma conclusion anterior

In [21]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index().head() 

2016-03-05    0.001075
2016-03-06    0.004323
2016-03-07    0.005419
2016-03-08    0.007342
2016-03-09    0.009638
Name: last_seen, dtype: float64

In [22]:
autos["last_seen"].str[:10].value_counts(normalize=True, dropna=False).sort_index(ascending = False ).head()

2016-04-07    0.132097
2016-04-06    0.221650
2016-04-05    0.124982
2016-04-04    0.024508
2016-04-03    0.025170
Name: last_seen, dtype: float64

Ahora se explorará "registration_year"

In [23]:
autos["registration_year"].describe()

count    48351.000000
mean      2004.780418
std         88.833591
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

Se observa que:
* tanto el maximo año como el mínimo son inusuales.  
* Puesto que la muestra es de 50.000 datos mediante count vemos que hay alguno datos Nulos. 

In [24]:
autos["registration_year"].sort_values(ascending = True ).head(20)

22316    1000
49283    1001
24511    1111
32585    1800
10556    1800
22659    1910
28693    1910
21416    1927
22101    1929
11246    1931
2221     1934
2573     1934
23804    1937
26607    1937
21421    1937
39725    1937
26103    1938
24855    1939
25792    1941
13963    1941
Name: registration_year, dtype: int64

In [25]:
autos["registration_year"].sort_values(ascending = False ).head(20)


8012     9999
33950    9999
38076    9999
49910    9000
25003    8888
8360     6200
27618    5911
4164     5000
49153    5000
22799    5000
24519    5000
42079    4800
453      4500
4549     4100
27578    2800
5763     2019
49185    2019
44256    2018
46811    2018
28335    2018
Name: registration_year, dtype: int64

Se observa que tanto en el límite inferior como en el superior en fechas existen, valores inconsistentes. Por un lado la máxima fecha posible es el año 2016 ( fecha de extraccion de la data) y el mínimo posible es 1901 (primer auto en Alemania). Por ellos se realizaran los siguientes pasos:

* Se analizara que porcentaje total de los datos corresponden las fechas fuera del límite 1995< x < 2016. Esto se realiza para ver si al pérdida de los mismos es significativa. 
* En caso de que no sea significativa se eliminará los datos de esas fechas.  

In [26]:
((~autos["registration_year"].between(1900,2016)).sum()/autos.shape[0])*100

3.8778929081094495

Observamos que corresponde solamente al 4% de nuestros datos, por ello se elimina ese rango de nuestro Data Frame.

In [32]:
autos = autos[(autos["registration_year"].between(1900,2016))]
autos["registration_year"].sort_values().head()

22659    1910
28693    1910
21416    1927
22101    1929
11246    1931
Name: registration_year, dtype: int64

In [34]:
autos["registration_year"].sort_values(ascending=False).head()

46998    2016
5607     2016
7171     2016
34850    2016
37852    2016
Name: registration_year, dtype: int64

### Análisis de la columna Brand

La idea es explorar las variaciones que ha sufrido una mima marca. Para ello se utiliza la técnica de agregación. Los pasos siguientes para raelizar tal técnica son:

- Identificar los valores únicos que se desean agregar
- Crear un diccionario vacio en el cual se incorporara los datos agregados
- Utilizar un loop por cada "unique" valor y con el:
    - Filtrar la Data Frame por el unique value.
    - Calcular el valor promedio de la columna de interes
    - Asignar el valor promedio junto con su key al diccionario

In [110]:
autos_top = autos["brand"].value_counts().head(5)  # se eligen las marcas top 20 para agregar
print(autos_top)

volkswagen       9824
bmw              5110
opel             4987
mercedes_benz    4485
audi             4025
Name: brand, dtype: int64


In [115]:
sel_brands = autos_top.index.unique()                    # se extraen los nombres de las marcas
brands_mean = {}                                        # diccionario vacio a utilizar
print(sel_brands)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi'], dtype='object')


In [117]:
for elem in sel_brands:
 selrows = autos[autos["brand"]== elem ]
 mean_price = selrows["price"].mean()
 brands_mean[elem] = int(mean_price)
print(brands_mean)

{'volkswagen': 5423, 'bmw': 8376, 'audi': 9373, 'mercedes_benz': 8663, 'opel': 2996}


In [122]:
print( "Diferencia de precion entre la mas popular y la quinta mas popular:",brands_mean['volkswagen'] - brands_mean['opel'])
print( "Diferencia de precion entre la mas popular y la mas cara:",brands_mean['audi'] - brands_mean['volkswagen'])

Diferencia de precion entre la mas popular y la quinta mas popular: 2427
Diferencia de precion entre la mas popular y la mas cara: 3950


* Audi, BMW and Mercedes Benz las mas caras
* Opel la mas barata
* Volkswagen en el medio


### Análisis de la columna Kilometros recorridos vs Precios

Se quiere saber si existe alguna relacion entre el precio y los kilometros recoridos por los autos. Para ello se utilizara una combinacion de marca-precio-odometro, creando una sola columna para tales propósitos, es decir se combinarán dos series de nuestro Data Frame. Los métodos a utilizar son: 
pandas series constructor y pandas dataframe constructor.

En primer lugar se debe obtener el diccionario que falta: Kilometros diccionario:


In [135]:
km_mean = {} 
for elem in sel_brands:
 selrows = autos[autos["brand"]== elem ]
 mean_km = selrows["odometer_km"].mean()
 km_mean[elem] = int(mean_km)
print(km_mean)

{'volkswagen': 128757, 'bmw': 132651, 'audi': 129254, 'mercedes_benz': 130923, 'opel': 129440}


Con esto ya se tiene marca-precio y marca-kilometros diccionarios. Ahora se procede a transformacion de estos diccionarios un dataFrame de una columna utilizando los panda seires y panda dataframe constructors. 

In [124]:
bmp_series = pd.Series(brands_mean) 
print(bmp_series) 

audi             9373
bmw              8376
mercedes_benz    8663
opel             2996
volkswagen       5423
dtype: int64


Como constructor (pandas series) crea un objeto de class pd.Series, donde transforma el diccionario en un objeto. La key en el diccionario se convierte en index y value en su valor en una columna, ahora falta darle nombre a la columna y para ello se utiliza el Data Frame constructor, el que crea un nuevo DataFrame de una columna.  

In [136]:
df = pd.DataFrame(bmp_series, columns=['price_mean'])
df

price_mean
audi                 9373
bmw                  8376
mercedes_benz        8663
opel                 2996
volkswagen           5423

In [137]:
km_series = pd.Series(km_mean) 
print(km_series) 

audi             129254
bmw              132651
mercedes_benz    130923
opel             129440
volkswagen       128757
dtype: int64


Agrenando la serie de kilometros al dataframe antes creado:

In [139]:
df["km_mean"] = km_series
df


price_mean  km_mean
audi                 9373   129254
bmw                  8376   132651
mercedes_benz        8663   130923
opel                 2996   129440
volkswagen           5423   128757

Finalmente se observa que el kilometrage por cada marca no varia demasiado a diferencia del precio en cada marca,lo que se puede suponer que en cuanto a rendimiento tanto los autos mas caros como los mas baratos permiten realizar el mismo recorrido,De esta forma, finalmente se mostrara una tabla de cuanto sale cada kilometro que se recorre por cada auto ( solo en cuanto a su precio, claro esta)


In [145]:
costo_km = df["km_mean"] / df["price_mean"]
print(costo_km)

audi             13.790035
bmw              15.837034
mercedes_benz    15.112894
opel             43.204272
volkswagen       23.742762
dtype: float64


Vemos que la marca Opel nos da mas de el doble de kilometros por cada euro que se paga por los mismos kilometros recorridos por los autos mas caros. 